middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml

```diff
@@ -0,0 +1,431 @@
+    | Infix_header ->
+      let field =
+        P.infix_header ~first_value_slot_offset:(slot_offset + 1) ~dbg
+      in
+      field :: acc, slot_offset + 1, env, Ece.pure, updates
+    | Value_slot v ->
+      let simple = Value_slot.Map.find v value_slots in
+      let contents, env, eff = P.simple env simple in
+      let env, fields, updates =
+        match contents with
+        | `Data fields -> env, fields, updates
+        | `Var v ->
+          (* We should only get here in the static allocation case. *)
+          assert (Option.is_some symbs);
+          let set_of_closures_symbol =
+            get_whole_closure_symbol ~set_of_closures_symbol_ref
+          in
+          let env, updates =
+            C.make_update env dbg Word_val
+              ~symbol:(C.symbol set_of_closures_symbol)
+              v ~index:slot_offset ~prev_updates:updates
+          in
+          env, [P.int ~dbg 1n], updates
+      in
+      List.rev fields @ acc, slot_offset + 1, env, eff, updates
+    | Function_slot c -> (
+      let code_id = Function_slot.Map.find c decls in
+      let code_linkage_name = Code_id.linkage_name code_id in
+      let arity, closure_code_pointers, dbg =
+        Env.get_func_decl_params_arity env code_id
+      in
+      let closure_info =
+        C.closure_info ~arity ~startenv:(startenv - slot_offset)
+      in
+      let acc =
+        match symbs with
+        | None -> acc
+        | Some symbs ->
+          let function_symbol = Function_slot.Map.find c symbs in
+          P.define_global_symbol (Symbol.linkage_name_as_string function_symbol)
+          @ acc
+      in
+      (* We build here the **reverse** list of fields for the function slot *)
+      match closure_code_pointers with
+      | Full_application_only ->
+        let acc =
+          P.int ~dbg closure_info
+          :: P.symbol_from_linkage_name ~dbg code_linkage_name
+          :: acc
+        in
+        acc, slot_offset + 2, env, Ece.pure, updates
+      | Full_and_partial_application ->
+        let acc =
+          P.symbol_from_linkage_name ~dbg code_linkage_name
+          :: P.int ~dbg closure_info
+          :: P.symbol_from_linkage_name ~dbg
+               (Linkage_name.create (C.curry_function_sym arity))
+          :: acc
+        in
+        acc, slot_offset + 3, env, Ece.pure, updates)
+
+  let rec fill_layout ~set_of_closures_symbol_ref symbs decls dbg ~startenv
+      value_slots env effs acc updates ~starting_offset slots =
+    match slots with
+    | [] -> List.rev acc, starting_offset, env, effs, updates
+    | (slot_offset, slot) :: slots ->
+      let acc =
+        if starting_offset > slot_offset
+        then
+          Misc.fatal_errorf "Starting offset %d is past slot offset %d"
+            starting_offset slot_offset
+        else if starting_offset = slot_offset
+        then acc
+        else
+          acc
+          @ List.init (slot_offset - starting_offset) (fun _ -> P.int ~dbg 1n)
```

Comment:
Shouldn't this concatenation be the other way ? I.e. put the dummy `1` elements at the front of the list, considering we build the reversed list of fields...